doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,697 @@
+     faster than ``CHPL_TARGET_MEM=cstdlib`` (see also
+     :ref:`readme-chplenv.CHPL_TASKS` and
+     :ref:`readme-chplenv.CHPL_TARGET_MEM`)
+
+   * ``CHPL_TARGET_COMPILER=llvm`` might or might not be faster than
+     using the C backend with something like
+     ``CHPL_TARGET_COMPILER=gnu``; however the performance of the LLVM
+     backend is more reliable (see also
+     :ref:`readme-chplenv.CHPL_COMPILER`)
+
+     * if you are using ``CHPL_LLVM=system``, it's a good idea to match the
+       version of LLVM bundled in the Chapel release if possible as this
+       has recieved the most attention and testing (see also
+       :ref:`readme-chplenv.CHPL_LLVM`)
+
+ * For multi-locale programs, use a high-performance networking configuration
+
+   * The following configurations support portability and correctness
+     testing but **are NOT designed for high performance**:
+
+     * ``CHPL_COMM=gasnet`` ``CHPL_COMM_SUBSTRATE=udp``
+
+     * ``CHPL_COMM=gasnet`` ``CHPL_COMM_SUBSTRATE=mpi``
+
+   * High-performance networking configuration will depend on your
+     system:
+
+     * For HPE Cray EX, use ``CHPL_COMM=ofi`` (see also :ref:`readme-cray`)
+     * For InfiniBand systems, use ``CHPL_COMM=gasnet``
+       ``CHPL_COMM_SUBSTRATE=ibv`` (see also :ref:`readme-infiniband`)
+     * For Omni-Path systems, use ``CHPL_COMM=gasnet`` and
+       ``CHPL_COMM_SUBSTRATE=ofi`` (see also :ref:`readme-omnipath`)
+
+
+Settings to Adjust to Improve Performance
+-----------------------------------------
+
+This section contains some easy things to try in order to improve
+performance.
+
+``--fast``
+  If you haven't been using ``--fast`` yet please do! It should be used
+  when measuring performance. Since it disables bounds checking, make
+  sure that your development flow includes correctness tests that aren't
+  compiled with ``--fast``.
+
+``--no-ieee-float`` / ``--ieee-float``
+  By default, only floating point optimizations that are relatively
+  benign are enabled. Depending on your application, you might use
+  ``--no-ieee-float`` to enable optimizations that might impact the
+  numerical accuracy. Or, if your program relies on floating point
+  operations happening in the order written for numerical accuracy, you
+  should use ``--ieee-float``.
+
+colocales
+  In some settings, running with multiple colocales per node can improve
+  performance. For example, to run on 8 nodes with 2 processes per node,
+  you could use ``-nl 8x2``. That will result in 16 locales; where each
+  node has 2 locales. Using colocales can help with memory bandwidth on
+  NUMA systems and also can better use the networking resources on a node
+  to help to make communication more efficient.
+
+``--auto-aggregation``
+  This compiler flag enables an optimization that automatically uses
+  aggregators to improve multilocale performance. It is not on by default
+  because it can slow down some applications.
```

Comment:
Maybe add

> This optimization can be beneficial if your application has small `forall` loops that do a lot of random remote accesses. If `forall` loops do mostly local accesses, it can hurt performance.